### PR TITLE
Pattern matcher: extractors become name-based.

### PR DESCRIPTION
**Pull request by paulp**
An extractor is no longer required to return Option[T], and
can instead return anything which directly contains methods
with these signatures:

```
  def isEmpty: Boolean
  def get: T
```

If the type of get contains methods with the names of
product selectors (_1, _2, etc.) then the type and arity
of the extraction is inferred from the type of get. If
it does not contain _1, then it is a single value
extractor analogous like Option[T].

This has significant benefits and opens new territory:
- an AnyVal based Option-like class can be used which
  leverages null as None, and no allocations are necessary
- for primitive types the benefit is squared (see below)
- the performance difference between case classes and
  extractors should now be largely eliminated
- this in turn allows us to recapture great swaths of
  memory which are currently squandered (e.g. every
  TypeRef has fields for pre and args, even though these
  are more than half the time NoPrefix and Nil)

Here is a primitive example:

```
  final class OptInt(val x: Int) extends AnyVal {
    def get: Int = x
    def isEmpty = x == Int.MinValue // or whatever is appropriate
  }
  // This boxes TWICE: Int => Integer => Some(Integer)
  def unapply(x: Int): Option[Int]
  // This boxes NONCE
  def unapply(x: Int): OptInt
```

As a multi-value example, after I contribute some methods to TypeRef:

```
  def isEmpty = false
  def get     = this
  def _1      = pre
  def _2      = sym
  def _3      = args
```

Then its extractor becomes

```
  def unapply(x: TypeRef) = x
```

Which, it need hardly be said, involves no allocations.

### CODE REVIEW

**Comment by paulp** [comment]
There's still probably plenty to quibble with, but this at least presents a basis for discussion.

**Comment by paulp** [comment]
I am assuming the failure in pr-scala-distpack is not on my conscience.

**Comment by adriaanm** [comment]
I'm a fan of the refactoring in d37e629, but could you please make a rough summary of behavior changes in that commit. I went through the codegen bit, but the behavior signal in match translation and typers is hard to separate from the refactoring "noise". Based on initial skimming, I expect a swift path to LGTM.

**Comment by paulp** [comment]
I will jump to the conclusion that the failures in pr-scala-integrate-partest aren't my doing either.

```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: org.scala-lang.modules#scala-xml_2.11.0-M4;1.0-RC2: not found
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
sbt.ResolveException: unresolved dependency: org.scala-lang.modules#scala-xml_2.11.0-M4;1.0-RC2: not found
    at sbt.IvyActions$.sbt$IvyActions$$resolve(IvyActions.scala:213)
    at sbt.IvyActions$$anonfun$update$1.apply(IvyActions.scala:122)
```

**Comment by paulp** [comment]
That last push accounts for 7214. I'm not going to bother with any please-rebuilds because I assume it will just fail again.

**Comment by adriaanm** [comment]
I went through the whole diff. Excellent work. Good to merge on Monday, as far as I'm concerned.

Your refactoring is a great improvement, but, for my understanding, is there anything in MatchTranslation that is crucial for the new translation?

Here are my Cliff's notes for understanding the new implementation of the new behavior:
- https://github.com/paulp/scala/commit/0b4a8320139f83197e318e5c4d9a27fdd0a02092#L4L135
- https://github.com/paulp/scala/commit/0b4a8320139f83197e318e5c4d9a27fdd0a02092#L8R177 - https://github.com/paulp/scala/commit/0b4a8320139f83197e318e5c4d9a27fdd0a02092#L8R209
- https://github.com/paulp/scala/commit/0b4a8320139f83197e318e5c4d9a27fdd0a02092#L2R463
- https://github.com/paulp/scala/commit/0b4a8320139f83197e318e5c4d9a27fdd0a02092#L11R155

Anything I missed?

File: src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+    }
+
+    case class UnapplyMethodInfo(unapply: Symbol, tpe: Type) {
+      def name         = unapply.name
+      def isUnapplySeq = name == nme.unapplySeq
+      def unapplyType  = tpe memberType method
+      def resultType   = tpe.finalResultType
+      def method       = unapplyMember(tpe)
+      def paramType    = firstParamType(unapplyType)
+      def rawGet       = if (isBool) UnitTpe else resultOfMatchingMethod(resultType, "get")()
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala:193]: How about `nme.get` here?

File: src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+      def rawTypes     = if (isBool) Nil else if (rawProduct.isEmpty) rawGet :: Nil else rawProduct
+      def rawArity     = rawTypes.size
+      def isBool       = resultType =:= BooleanTpe   // aka "Tuple0" or "Option[Unit]"
+      def isCase       = method.isCase
+
+      private def rawProduct = getNameBasedProductSelectorTypes(rawGet)
+    }
+
+    object NoUnapplyMethodInfo extends UnapplyMethodInfo(NoSymbol, NoType) {
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala:202]: Used anywhere?

**Comment by xeno-by** [comment]
Gone are the days of oppression. Unapply macros are finally making it into trunk!

```
import scala.reflect.macros.Context
import language.experimental.macros

trait Tree
case object SomeTree extends Tree

object NewQuasiquotes {
  implicit class QuasiquoteInterpolation(c: StringContext) {
    object nq {
      def unapply(t: Tree) = macro QuasiquoteMacros.unapplyImpl
    }
  }
}

object QuasiquoteMacros {
  def unapplyImpl(c: Context)(t: c.Tree) = {
    import c.universe._
    q"""
      new {
        def isEmpty = false
        def get = this
        def _1 = SomeTree
        def _2 = SomeTree
        def unapply(t: Tree) = this
      }.unapply($t)
    """
  }
}

object Test extends App {
  import NewQuasiquotes._
  SomeTree match {
    case nq"$x + $y" => println((x, y))
  }
}

09:10 ~/Projects/patmat/sandbox (topic/patmat)$ ssr
(SomeTree,SomeTree)
```

As brought up by Denys yesterday, this pull request will finally allow us to turn unapplications into macros without resorting to insane hacks like manual assembly of UnApply nodes. Oh yeah!

**Comment by retronym** [comment]
Could you submit a follow up pr with a few test cases for unapply macros? Would be great to nail them down. 

IntelliJ will also need to be updated with the relaxed rules for extractors. SIP worthy? /cc @alefas

**Comment by xeno-by** [comment]
@retronym Yes, that's what I'm doing right now.

**Comment by xeno-by** [comment]
@retronym and all: https://github.com/paulp/scala/pull/8 establishes a pattern that can be used to implement extractor macros, explains how to implement it and provides examples in tests.

**Comment by xeno-by** [comment]
@paulp Btw when playing with extractor macros, I noticed a gotcha:

```
object Matcher {
  def isEmpty = false
  def get = this
  def _1 = 2
  def unapply(x: Int) = this
  override def toString = "oops"
}

2 match {
  case Matcher(x) => println(x) // will print oops!
}
```

Is this intended?

**Comment by paulp** [comment]
@xeno-by this has been bugging me and I'm not sure what to do about it. I was going to bring it up today. It's only an issue with Product1 (or something else which defines _1 and not _2 for whatever reason) but in such cases I don't know what the right thing is to do. I'm surprised it prints oops, I thought the logic would have it printing 2, but I'm not sure 2 isn't a problem also.

**Comment by paulp** [comment]
@adriaanm I got stuck in some other stuff yesterday, I will elaborate/comment today.

**Comment by dragos** [comment]
It seems the [IDE validation fails](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1398/consoleFull) with a compilation error: (the Jenkins job looks blue, though):

```
[ERROR] /localhome/jenkins/c/workspace/pr-scala-integrate-ide/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/formatter/ScalaFormatterPreferenceInitializer.scala:20: error: overloaded method value setDefault with alternatives:
[ERROR]   (x$1: String,x$2: Boolean)Unit <and>
[ERROR]   (x$1: String,x$2: String)Unit <and>
[ERROR]   (x$1: String,x$2: Long)Unit <and>
[ERROR]   (x$1: String,x$2: Int)Unit <and>
[ERROR]   (x$1: String,x$2: Float)Unit <and>
[ERROR]   (x$1: String,x$2: Double)Unit
[ERROR]  cannot be applied to (String, _66)
[ERROR]           preferenceStore.setDefault(preference.eclipseKey, pd.defaultValue)
```

**Comment by adriaanm** [comment]
We had to make the jenkins job always succeed because the build flow plugin's ignore combinator is broken.

**Comment by paulp** [comment]
@dragos thanks for pointing it out, that looks like a error which needs addressing here.

**Comment by dragos** [comment]
@paulp for the record, the overloaded method is defined in a trait hierarchy in [PreferenceDescriptor.scala](https://github.com/mdr/scalariform/blob/master/scalariform/src/main/scala/scalariform/formatter/preferences/PreferenceDescriptor.scala)

**Comment by paulp** [comment]
I'm sure I'll be able to fix it. This is one of those brutal things where some arbitrary overloading resolution is done in-place on the tree (directly modifying the symbol and type in the midst of typing) so of course no matter how one touches it something somewhere will break.

**Comment by adriaanm** [comment]
Here are some other failures (failing to compile scalariform):
- 0b4a8320139f83197e318e5c4d9a27fdd0a02092: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1404
- bef86df163f817861f56502acaa9145de941de31: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1403
- 7916a7397ed99df08b3c2c4aac1ae4268dc6bca7: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1353
- 912a1c32558b5275166a3c3aeb3887ce6ac43f10: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1352
- 43455eac97146a65b9835bc242e09de2010efe86: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1341
- 2ef2c2d8a42204ac8527333bf44365cb24e24ac4: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1339
- 2ef2c2d8a42204ac8527333bf44365cb24e24ac4: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1338
- 34e29721ff508572fa55e4e4254cc11ea81fa2ac: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1313
- 6b7213c7096cfd8f72f9ce6e5528c7ad8b85b5d3: https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1311

@dragos, while looking at the IDE logs I noticed some new failures: org.scala-ide.sbt.full.library, org.scala-ide.sdt.core, org.scala-ide.sdt.core.tests, org.scala-ide.sdt.debug.tests, org.scala-refactoring.library, scalariform (the list above), Sbinary build failures on 2.10 (it used to work, but now complains about xml) -- I haven't looked into them.

**Comment by dragos** [comment]
@adriaanm I'm going through the most recent.. Do you have a tip on how to find only failures? I'm clicking my way through them, but it's very time consuming.

**Comment by paulp** [comment]
See my comment in #2848 ; there was definitely a bug on my end which was shown in scalariform, but I don't know to what extent that will explain the current crop of failures.

**Comment by paulp** [comment]
Anywhere you see this:

```
[ERROR] error: symbol value key does not exist in scalariform.formatter.preferences.PreferencesImporterExporter.getPreferences
[ERROR] error: scala.reflect.internal.FatalError: symbol value key does not exist in scalariform.formatter.preferences.PreferencesImporterExporter.getPreferences
```

That's my bug.

**Comment by adriaanm** [comment]
@dragos: The script does the grepping for you, should be extensible to find other failures. To clarify: the script I attached in that thread. I've gisted it: https://gist.github.com/adriaanm/6258066

**Comment by paulp** [comment]
Superseded by #2848.

**Comment by paulp** [comment]
@dragos I've reproduced the overload issue, hopefully that will be the hard part. I'm sort of impressed we have no test which exercises this. It doesn't even have to be overloaded.

```
trait Bip[T] { def h: T }
trait BoolBip extends Bip[Boolean]

class A {
  def g(x: Boolean): Unit = ()
  def f(xs: List[Bip[_]]) = xs foreach { case x: BoolBip => g(x.h) }
}
```

```
b.scala:6: error: type mismatch;
 found   : _1
 required: Boolean
  def f(xs: List[Bip[_]]) = xs foreach { case x: BoolBip => g(x.h) }
                                                                ^
one error found
```

**Pull request closed without merging**